ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -158,14 +259,27 @@ def distances_to_end(self, waypoints):
             incremental_dist = dl(
                 waypoints[wp_idx].pose.pose.position,
                 waypoints[wp_idx + 1].pose.pose.position)
-
+            # Accumulate previous distance upto this waypoint and add the current distance
             total_dist = dists_reversed[-1] + incremental_dist
 
             dists_reversed.append(total_dist)
-
+        # reverse the list so that it contains waypoints distance starting from closest waypoint
         return dists_reversed[::-1]
 
 
+    def get_waypoint_velocity(self, waypoint):
+
+        # Get velocity for a given waypoint object
+        return waypoint.twist.twist.linear.x
+
+    def set_waypoint_velocity(self, waypoints, waypoint, velocity):
+
+        # Set velocity for a given waypoint in the list of waypoints
+        waypoints[waypoint].twist.twist.linear.x = velocity
```

Comment:
We definitely do NOT want to override waypoints, which are supposed to be immutable. I remember the walkthrough videos mentioned this.

The waypoints contain the reference speed, yaw, etc for each point (and they come from `wp_yaw_const.csv`). Because we modify them, we lose these references immediately.

This may not matter in practice if we can set a constant speed target for every waypoint. But it may cause other problems in unknown ways, especially in Udcacity's backyard.

Please modify your algorithm to build path in a separate array than waypoints.

Reply:
We get the waypoints once only.

Reply:
I only modify velocity and nothing else in the original waypoints... I intentionally kept it that way so we have the freedom to modify the reference velocity.. I don't think it would matter in practice but yes I think you are right, it might cause problems in Udacity's backyard. I will take the reference velocity from waypoints. Btw is there any upper limit for the reference speed? I can't exactly remember if there was a limitation to max speed.

Reply:
The constant reference speed is the target speed at all waypoints.

The bottom line is we need to have two constant reference speeds: one for simulation and one for real world. You can get it from something like `rospy.get_param('~velocity')`.